helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -486,187 +577,273 @@ public void updateDeploymentGroupHosts(final String groupName, final List<String
+    for (final String id : existing.subList(MAX_ROLLING_OPERATION_HISTORY - 1, existing.size())) {
+      removeRollingOperation(client, operations, id);
     }
 
+    return existing.subList(0, MAX_ROLLING_OPERATION_HISTORY - 1);
+  }
+
+  private void prependRollingOperation(final ZooKeeperClient client,
+                                       final List<ZooKeeperOperation> operations,
+                                       final String groupName,
+                                       final String rollingOpId)
+      throws DeploymentGroupDoesNotExistException {
+    final List<String> rollingOps = new ArrayList<String>();
+    rollingOps.add(rollingOpId);
+    try {
+      try {
+        // addDeploymentGroup creates Paths.statusDeploymentGroupRollingOps(name) so it should
+        // always exist. If it doesn't, then the DG was (likely) deleted.
+        final Node node = client.getNode(Paths.statusDeploymentGroupRollingOps(groupName));
+        final List<String> all = Json.read(node.getBytes(), new TypeReference<List<String>>() {});
+        rollingOps.addAll(truncateRollingOps(client, operations, all));
+      } catch (NoNodeException e) {
+        throw new DeploymentGroupDoesNotExistException(groupName, e);
+      } catch (JsonMappingException e) {
+        // Can't parse JSON, just replace it with the fresh list of one rollingOpName.
+      }
+      operations.add(set(Paths.statusDeploymentGroupRollingOps(groupName),
+                         Json.asBytes(rollingOps)));
+    } catch (KeeperException | IOException e) {
+      throw new HeliosRuntimeException("updating deployment group rolling-operations failed", e);
+    }
+  }
+
+  private RollingUpdateOp generateOps(final ZooKeeperClient client, final RollingOperation rolling)
+      throws DeploymentGroupDoesNotExistException {
     final List<ZooKeeperOperation> operations = Lists.newArrayList();
-    final ZooKeeperClient client = provider.get("rollingUpdate");
+    final List<String> hosts = getDeploymentGroupHosts(rolling.getDeploymentGroupName());
+    return generateOps(client, rolling, operations, hosts);
+  }
+
+  private RollingUpdateOp generateOps(final ZooKeeperClient client,
+                                      final RollingOperation rolling,
+                                      final List<ZooKeeperOperation> operations,
+                                      final List<String> hosts)
+    throws DeploymentGroupDoesNotExistException {
+    final RollingOperationStatus status = RollingOperationStatus.newBuilder()
+        .setState(NEW)
+        .build();
 
-    operations.add(set(Paths.configDeploymentGroup(updated.getName()), updated));
 
     try {
-      final RollingUpdateOp op = getInitRollingUpdateOps(updated, MANUAL, client);
-      operations.addAll(op.operations());
+      // Fail early if the deployment group does not exist.
+      if (client.exists(Paths.configDeploymentGroup(rolling.getDeploymentGroupName())) == null) {
```

Comment:
Most likely it just didn't occur to me at the time. Though possibly it was to reuse the existing `ZookeeperClient`. I suppose I could overload `getDeploymentGroup` to allow passing in a client.